### ISSUE TITLE
Cygfuse.pc should be named fuse.pc

### ISSUE THREAD

**Issue by billziss-gh** [mgeisert/cygfuse#1]
I believe that `cygfuse.pc` should be renamed `fuse.pc` so that build configuration scripts can easily find it. For example, SSHFS looks for `fuse >= 2.3` is its `configure.ac`:

```
PKG_CHECK_MODULES([SSHFS], [fuse >= 2.3 glib-2.0 gthread-2.0])
```

An additional change in `cygfuse.pc` is that its `Description` should have no references to WinFsp. Perhaps use `"FUSE compatibility layer"` (same as the `SUMMARY` in `cygfuse.port`).

**Comment by mgeisert** [comment]
Agreed. cygfuse.pc.in will be renamed to fuse.pc.in and updated as you suggest. Makefile will be updated to refer to the renamed file. Should I also change the URL value in fuse.pc.in to refer to this GitHub project?

**Comment by billziss-gh** [comment]
> Should I also change the URL value in fuse.pc.in to refer to this GitHub project?

I agree.

**Comment by mgeisert** [comment]
Committed. Thanks for the feedback!

**Comment by Rondom** [comment]
I disagree that hijacking the fuse-name from the fuse-project is a good idea. libfuse and cygfuse are two completely independent projects and in my opinion the configure-scripts should be changed to accept either library.
There are other precedents, like libcurses and pdcurses which can mostly act as a drop-in replacement for the other, yet they can be installed alongside another and configure-scripts call pkg-config for each of the two names. In case a project does not consider all possible libraries that can be used as drop-in replacements, one can override it on command-line.

There has been an experimental implementation of the Linux FUSE Kernel API as Windows driver. While this is unmaintained and buggy as hell, it proves that running libfuse on Windows is not something that is "impossible".

**Comment by mgeisert** [comment]
What's intended here is different from the libcurses vs pdcurses example. In that example you install one or the other, and each is compatible with the other and with apps that require a curses package. Here, we intend to allow multiple FUSE-providing projects installed at the same time, with run-time selection by environment variable (unless something better is suggested). An app that requires FUSE shouldn't have to care which particular FUSE is installed, therefore we use 'fuse' as the generic name that config scripts should look for. App config scripts don't need to be modified, if this all goes as intended.

**Comment by Rondom** [comment]
I see your point and on the one hand I find it indeed convincing, but then on the other hand I would prefer to leave all options open, just in case.

First of all, there will always be case where one wants to directly link to some fuse-library. For example, you are not covering the whole libfuse-API (e.g. the lowlevel API is missing at the moment) and even if you were always doing that and keeping all wrappers up to date I as a user might still be interested to make the informed choice to link against libfuse should it ever become available for Windows. The issue I see also is that this is a different codebase claiming it is libfuse, while it is actually not and one might want to avoid the wrapper. Or one wants to avoid cygfuse because it does not guarantee license compliance (unfortunately both Dokan (LGPLv3 or later) and WinFSP (AGPLv3) have licenses that are sometimes difficult to deal with because they are not always compatible with some licenses such as GPLv2-only. Thus there are situations where one might prefer one or the other (and there could theoretically be libfuse as a third possibility), even if it is just to make the legal department happy.

So to summarize:
- the issue I see is on one hand leaving options open on the technical side
- and on the other hand political in terms of not hijacking the libfuse pkg-config name from Nikolaus and Miklos from libfuse.

What about getting some input, by sending a post to the libfuse mailing list (maybe with cygwin in CC)? Maybe I am just putting to much thought into this and libfuse-people are totally okay with this?

Correction:

```
I stand corrected, I have just double-checked and the project that implemented FUSE on Windows NT 4.0 I was referring to did not implement the kernel protocol, but the user level API. Still I think a fork/implementation of libfuse for Windows is not some totally contrived situation.)
```

**Comment by billziss-gh** [comment]
The point is to make it painless for FUSE file systems to be ported to Cygwin. Ideally you should just download the source and compile. You should not have to fiddle with configuration scripts, PKG_CONFIG_PATH variables, etc.

You are saying that in the name of "leaving options" the package should be renamed, thus creating frustration for everyone who tries to port a FUSE file system to Cygwin. Furthermore a patch/PR will be now required for every file system even if it is simply to fiddle with the PKG_CHECK_MODULES line.

> For example, you are not covering the whole libfuse-API (e.g. the lowlevel API is missing at the moment) and even if you were always doing that and keeping all wrappers up to date I as a user might still be interested to make the informed choice to link against libfuse should it ever become available for Windows.

The Windows file system architecture is very different from the VFS model which FUSE uses. One fundamental difference is that FUSE lowlevel does lookups by (parent_dir_inode, name) tuples, whereas Windows uses the full path everywhere (even in kernel-mode). This and other reasons make the FUSE lowlevel API ill-suited for Windows. Implementing it is possible of course, but it would effectively be an emulation and not how Windows really works.

> Or one wants to avoid cygfuse because it does not guarantee license compliance (unfortunately both Dokan (LGPLv3 or later) and WinFSP (AGPLv3) have licenses that are sometimes difficult to deal with because they are not always compatible with some licenses such as GPLv2-only.

FSF itself has moved to GPLv3 (and AGPLv3) while they recognize the incompatibility with GPLv2-only.

Are there really that many FUSE file systems under the GPLv2-only? Most GPLv2 licenses are really GPLv2-plus, which makes them automatically compatible with GPLv3.

> and on the other hand political in terms of not hijacking the libfuse pkg-config name from Nikolaus and Miklos from libfuse.

If libfuse produces their own Cygwin package they can easily provide their own fuse.pc. The situtation would be similar to the libcurses vs pdcurses example that you cited earlier.

> I think a fork/implementation of libfuse for Windows is not some totally contrived situation.

Dokany/CBFS/WinFsp is it. It is absurd to try to recreate the design of the FUSE kernel in Windows. Is it possible? Yes. Would it be a good design for Windows? No.

**Comment by Rondom** [comment]
You nearly have me convinced ;-)

> Dokany/CBFS/WinFsp is it. It is absurd to try to recreate the design of the FUSE kernel in Windows. Is it possible? Yes. Would it be a good design for Windows? No.

Indeed it would be a huge effort that would maybe not have many benefits. I was confusing different experimental implementations, bot of them being called FUSE-NT. There is one FUSE-NT that ran a loopback-CIFS under Windows NT 4.0 (paper) and then there is another FUSE-NT that uses libfuse on the userspace side @cemeyer/FUSE-NT which seems to be failed/unfinished attempt. I do not know whether it worked at all, but my memories were correct.

Anyways, my main concern is a combination of respecting libfuse and an uneasy feeling about license compliance. Furthermore, I have an uneasy feeling about taking names of other projects without consulting with them first, so I would personally ask the libfuse people.

**Comment by billziss-gh** [comment]
> Anyways, my main concern is a combination of respecting libfuse and an uneasy feeling about license compliance. Furthermore, I have an uneasy feeling about taking names of other projects without consulting with them first, so I would personally ask the libfuse people.

The package name is CYGFUSE on GitHub and it will be the same on the Cygwin package list. We only use the name fuse.pc in one file that is installed in /usr/lib/pkgconfig. This is for convenience/compatibility only and not to confuse anyone.

The practice of using a name of {Platform}FUSE is widespread. See, for example, Macfuse, OSXFUSE and even FUSE-NT that you cited.

**Comment by Rondom** [comment]
> The package name is CYGFUSE on GitHub and it will be the same on the Cygwin package list

No disagreement about that. 

> We only use the name fuse.pc in one file that is installed in /usr/lib/pkgconfig.

And I was explicitly referring to the pkg-config.

The userspace libraries of the example you mentioned are all ports/forks of libfuse. Whether that makes a difference... is debatable. :-) For me it makes a small difference (it is not a completely different codebase like cygfuse), but discussing this deviates from the actual issue. Following my argument, one might even argue that each of those should have a different pkg-config-name because they are forks that have no immediate intention to be merged back to the main repository. And having a different name /might/ avoid people from showing up on the @libfuse bugtracker looking for osxfuse-support like e.g. in libfuse/libfuse#76, (but then again I am not claiming that it does.

> This is for convenience/compatibility only and not to confuse anyone.

Convenience is nice and even before hearing your arguments I could see the appeal of it.

I think everything has been said here to the point that I am repeating myself and I do not care too much about this to be honest. I think it is just, that if I was in the  libfuse authors' shoes I would prefer to be asked.

So to put it short: If the libfuse-people agree, you got me convinced. But seriously, no one says that you have to convince me :-) Still, I like to be convinced...

**Comment by billziss-gh** [comment]
> So to put it short: If the libfuse-people agree, you got me convinced. But seriously, no one says that you have to convince me :-) Still, I like to be convinced...

I am beginning to think that we cannot convince you.

I would respectfully ask however to not create additional work for us. If every small decision has to be discussed with everyone then no work will ever get done.

**Comment by Rondom** [comment]
> I am beginning to think that we cannot convince you.

That's fine and as I said I am 90% convinced now :-)

> I would respectfully ask however to not create additional work for us

Did I create that impression? To be clear, my intention was not "to create additional work". My interest was/is to have the cleanest solution and mainly respecting licenses of projects and the people behind them to the benefit of everyone involved.

**Comment by mgeisert** [comment]
I appreciate the discussion. I had assumed that 'fuse' was a generic identifier for package selection and I did not know it was already being used by another project, the definitive-for-Linux libfuse project at that. I've been kicking around a possible solution where cygfuse's fuse.pc is renamed in source to cygfuse.pc, but when installed on Cygwin a symlink is made to fuse.pc so that either name works when configuring packages on Cygwin. If/when the libfuse folks ship something for Cygwin, we can deal with the fuse.pc name collision. Open question is whether this change would bork a Windows-native install of WinFSP.

**Comment by billziss-gh** [comment]
> Open question is whether this change would bork a Windows-native install of WinFSP.

I do not expect any problems for WinFsp. When using Cygfuse WinFsp is not even aware of the fuse.pc file.
